frontends/concrete-python/concrete/fhe/mlir/processors/assign_bit_widths.py

```diff
@@ -13,266 +13,346 @@
 
 class AssignBitWidths(GraphProcessor):
     """
-    Assign a precision to all nodes inputs/output.
-
-    The precisions are compatible graph constraints and MLIR.
-    There are two modes:
-        - single precision: where all encrypted values have the same precision.
-        - multi precision: where encrypted values can have different precisions.
+    AssignBitWidths graph processor, to assign proper bit-widths to be compatible with FHE.
     """
 
     def __init__(self, single_precision=False):
         self.single_precision = single_precision
 
     def apply(self, graph: Graph):
-        nodes = graph.query_nodes()
-        for node in nodes:
-            assert isinstance(node.output.dtype, Integer)
-            node.properties["original_bit_width"] = node.output.dtype.bit_width
+        solver = z3.Solver()
 
-        if self.single_precision:
-            assign_single_precision(nodes)
-        else:
-            assign_multi_precision(graph, nodes)
+        bit_widths: Dict[Node, z3.Int] = {}
 
+        additional_constraints = AdditionalConstraints(solver, graph, bit_widths)
```

Comment:
could single precision done here (giving single precision as a parameter) ?
Imagine a multi-precision rule not compatible with the equal constraint after.

Reply:
No, we only have lover bound constraints as we are trying to minimize the bit-widths in the system.